### PR TITLE
NodalConstraints should support different variables on primary and secondary node sets

### LINKED ISSUE #21881: NodalConstraints should support different variables on primary and secondary node sets
## Reason
Some nodal constraints may constraint different variables on the primary and secondary node sets

## Design
Simple extension of `NodalConstraint` to account (optionally) for a different variable on the secondary node set

## Impact
Extended functionality.

### PR DESCRIPTION
**Pull request by matthiasneuner**
closes #21881

### CODE REVIEW

**Review by joshuahansel — CHANGES_REQUESTED** [review]
I believe you also need to make some changes in the following derived classes (in modules): 

- `NodalFrictionalConstraint`
- `NodalStickConstraint`

Alternatively, you could add error checks to them that `variable` equals `variable_secondary`.

**Comment by matthiasneuner** [comment]
> I believe you also need to make some changes in the following derived classes (in modules):
> 
>     * `NodalFrictionalConstraint`
> 
>     * `NodalStickConstraint`
> 
> 
> Alternatively, you could add error checks to them that `variable` equals `variable_secondary`.

I agree! I decided for the error, since I am not sure if different variables makes sense for those two constraints.

File: modules/tensor_mechanics/src/constraints/NodalFrictionalConstraint.C [hunk 1]
```diff
@@ -51,6 +51,10 @@ NodalFrictionalConstraint::NodalFrictionalConstraint(const InputParameters & par
     _u_secondary_old(_var.dofValuesOldNeighbor()),
     _u_primary_old(_var.dofValuesOld())
 {
+  if (_var.number() != _var_secondary.number())
+    mooseError("NodalFrictionalConstraint: Primary variable must be identical to secondary "
+               "variabe. Different variables are currently not supported.");
```
Comment (joshuahansel) [inline modules/tensor_mechanics/src/constraints/NodalFrictionalConstraint.C:56]: typo: variable. Same with other message.

**Review by joshuahansel — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/src/constraints/NodalFrictionalConstraint.C (hunk 1 above)
Comment (matthiasneuner) [reply modules/tensor_mechanics/src/constraints/NodalFrictionalConstraint.C]: Embarassing! Fixed that.

**Review by loganharbour — COMMENTED** [review]
It's not necessary to include the type of the object in `mooseError`. `MooseObject::mooseError()` adds a prefix to the error containing the object type and the name of the object.

See https://github.com/idaholab/moose/blob/next/framework/src/base/MooseObject.C#L77

**Comment by loganharbour** [comment]
And to add to the above... you should probably use `paramError()` instead of `mooseError()` in the cases you're erroring here.

**Comment by matthiasneuner** [comment]
> And to add to the above... you should probably use `paramError()` instead of `mooseError()` in the cases you're erroring here.

I see! I made it consistent with the existing message in `NodalFrictionalConstraint`, but at that time the prefix was probably not yet added. Fixed that for the new messages.

**Comment by matthiasneuner** [comment]
> It's not necessary to include the type of the object in `mooseError`. `MooseObject::mooseError()` adds a prefix to the error containing the object type and the name of the object.
> 
> See https://github.com/idaholab/moose/blob/next/framework/src/base/MooseObject.C#L77


I see! I made it consistent with the existing message in `NodalFrictionalConstraint`, but at that time the prefix was probably not yet added. Fixed that for the new messages. Oh,

**Review by joshuahansel — APPROVED** [review]

**Comment by matthiasneuner** [comment]
May I please ask again for the review @loganharbour :) ? Thank you!

**Review by loganharbour — APPROVED** [review]
👍🏼 going to re-run tests before merging as it's been a while

**Comment by loganharbour** [comment]
> > 👍🏼 going to re-run tests before merging as it's been a while
> 
> okay :)

Thanks for the reminder 😬

**Pull request merged**